docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,281 @@
+### Use cases
+After gathering client's expectations and a couple of brainstorming sessions, we ended up with multiple use cases for locality aware load balancing:
+
+1. Keeping traffic in single zone
+2. Load balancing based on priorities within a zone. Example: node → AZ → region
+3. Prefer traffic in single zone with fallback to other zones
+4. Load balance equally to list of zones
+5. When leaving zone, target specific zones first. Example zones physically close to each other
+6. Load balance equally to local dpp and other zone
+7. Exclude zone from load balancing
+
+### API design
+Looking at use cases, it is clear that we have two types of load balancing strategies, inside zone and cross zone. We can easily route to services with proper tags inside the zone, but immediately after we cross zone border and hit Zone Ingress we lose the ability to control locality.
+Because of this we have decided that we will split our API in two sections: `inZone` and `crossZone`.
+
+Here is the full example of policy:
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAware:
+      inZone: ["k8s.io/node", "k8s.io/az"] # (1)
+      crossZone: # (2)
+        rules: # (3)
+          - groups: # (4)
+            - ["zone-1", "zone-2"]
+            - ["zone-3", "zone-4"]
+          - groups:
+            - ["zone-5"]
+        fallback: any # (5)
+```
+
+
+(1) In `inZone` section, you configure the priorities of tags used for load balancing.
+
+(2) In `crossZone` section, you configure zones priority.
+
+(3) In `rules` section, you configure list of zone groups priorities, first groups will have the highest load balancing priority and last in order will have the lowest priority.
+
+(4) In `groups` section, you configure logical zone groups. Let's assume `zone-1` and `zone-2` are located in the same datacenter, so you want to load balance equally between them. `Zone-5` is located in another datacenter so it should have lower priority, that is why it is in the second groups entry.
+
+(5) At last we should apply some feedback if no dataplanes from previous configuration are available, at this moment we allow `fail` and `any` fallback
+
+### API examples based on use cases
+
+#### Keeping traffic in single zone
+
+![Use case 1](assets/031/use_case_1.png)
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAware:
+      inZone: []
+```
+
+#### Load balancing based on priorities within a zone. Example: node → AZ → region
+
+![Use case 2](assets/031/use_case_2.png)
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAware:
+      inZone: ["k8s.io/node", "k8s.io/az"]
+```
+
+#### Prefer traffic in single zone with fallback to other zones
+
+![Use case 3](assets/031/use_case_3.png)
```

Comment:
The drawings are very helpful though I think you should show all the cases for each of the DPP

Reply:
I was thinking about this, but I was afraid that it will get messy and harder to actually understand

Reply:
I think wiht your color codes it should be fine

Reply:
I've added separate diagram for each ddp to make it clear